File: lessons/bitmasking/bitmasking.c
#include <stdint.h>
#include <stdio.h>

enum Permission {
    PERM_READ = 1u << 0,
    PERM_WRITE = 1u << 1,
    PERM_EXEC = 1u << 2,
};

int main(void) {
    uint32_t perm = 0;
    perm |= PERM_READ | PERM_WRITE;
    printf("Set READ+WRITE -> 0x%X\n", perm);

    perm &= ~PERM_WRITE;
    printf("Clear WRITE -> 0x%X\n", perm);

    perm ^= PERM_EXEC;
    printf("Toggle EXEC -> 0x%X\n", perm);

    printf("Readable? %d\n", (perm & PERM_READ) != 0);
    printf("Writable? %d\n", (perm & PERM_WRITE) != 0);
    printf("Executable? %d\n", (perm & PERM_EXEC) != 0);
    return 0;
}

File: lessons/bitmasking/bitmasking.py
READ, WRITE, EXECUTE = 1 << 0, 1 << 1, 1 << 2

perm = 0
perm |= READ | EXECUTE
print(f"Set READ+EXEC -> {perm:03b}")

perm &= ~EXECUTE
print(f"Clear EXECUTE -> {perm:03b}")

perm ^= WRITE
print(f"Toggle WRITE -> {perm:03b}")

print("Readable?", bool(perm & READ))
print("Writable?", bool(perm & WRITE))
print("Executable?", bool(perm & EXECUTE))
